stdlib/public/runtime/Exclusivity.cpp

```diff
@@ -100,52 +121,120 @@ class AccessSet {
   }
 };
 
-} // end anonymous namespace
-
 enum {
   NumAccessSetBits = 2,
-  NumAccessSets = 1 << NumAccessSetBits
+  NumAccessSets = 1 << NumAccessSetBits,
+  NumPointerBits = (8 * sizeof(uintptr_t))
 };
 
-namespace {
-  template <unsigned K, bool Exceeds = (K >= (8 * sizeof(uintptr_t)))>
-  struct KBitHasher;
-
-  template <unsigned K>
-  struct KBitHasher<K, true> {
-    static uintptr_t hash(uintptr_t value) {
-      return value;
-    }
-  };
+/// A template for turning an integer into a hash with meaningful
+/// entropy in the bottom K bits.
+template <unsigned K, class T, bool = (K >= 8 * sizeof(T))>
+struct KBitHasher;
```

Comment:
Question, is this a hand rolled hash implementation? If so, why not just use what LLVM has for hashing (i.e. hash_combine and friends)? I don't think it has a dependency on any binary support (i.e. pure header). But maybe my memory is wrong.